enrolment-service/enrol/core/views.py

```diff
@@ -225,6 +225,67 @@ def perform_update(self, serializer):
             send_update_notification(email)
             # print(email) # for testing
 
+    def destroy(self, request, pk=None):
+        user = self.request.user
+
+        if user.is_superuser:
+            logger.debug('Attempting to purge project %s', pk)
+
+            try:
+            # We have to check for billing contacts separately, since it cannot cascade
+                logger.debug('Removing Billing Contact, if available')
+                billingContactPurgeResults = None
+
+                try:
+                    application = Applications.objects.get(project=pk)
+
+                except Exception as e:
+                    logger.debug('This project is not connected to an application (aka Orphaned): %s', e)
+
+                else:
+                    try:
+                        billingContact = application.billing_contact
+
+                        if billingContact != None:
+                            logger.debug('Found a Billing Contact. Purging it...')
+                            billingContactPurgeResults = billingContact.delete()
+
+                        else:
+                            logger.debug('No Billing Contact found. Moving on...')
+
+                    except Exception as e:
+                        logger.debug('Something wrong getting Billing Contact: %s', e)
+
+            # This is where the project itself is removed, along with its dependencies
+                finally:
+                    logger.debug('Purging project now...')
+                    projectPurgeResults = Projects.objects.get(id=pk).delete()
```

Comment:
That one is caught by the exception handler in line `282`, which returns a 500 for now.

Thought of making the response a 504, but then wondered "if the project wasn't found why was it even listed in the UI", which means something may actually be off and should be looked into in detail.
Does that make sense?

Reply:
error occurs because of user inputted parameter, thus is 400. even though the request originates form our UI and our UI shouldnt give them this option, thats the rule.

Reply:
this is a specific known failure, not a "something went wrong". 

should be handled separately as a "the ID you provided doesn't exist"

Reply:
🤔 hmm yeah, that tracks. Figured a data driven UI wouldn't have a pk for a non-existing project, so it must've been a server issue providing inaccurate data... but you're right in that the UI may fall out of sync on its own (or be tamprede with) and that's a client issue. Amending that!

Making it a 404 though, as it is technically a "records not found by the ID provided".
Do you agree?

Reply:
Ok, I see your point.

It did just dawn on me this exception can be triggered by anything else going wrong in that routine, so 500 suits; and an additional check is indeed missing to handle no project being found specifically.
Will add that instead, and make it a 400 there.

🙌